versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -757,7 +761,21 @@ protected CommitLogEntry buildIndividualCommit(
             Collections.emptyList());
 
     if (keyListDistance >= config.getKeyListDistance()) {
-      entry = buildKeyList(ctx, entry, newKeyLists);
+      Hash startHash;
+      if (unwrittenEntriesForKeyList.isEmpty()) {
+        // query all commits from backend.
+        startHash = entry.getParents().get(0);
+      } else if (entry.getParents().size() > unwrittenEntriesForKeyList.size()) {
```

Comment:
This comparison is wrong. the number of parent commits in a commit-entry should always be equal to  config.getParentsPerCommit(). I do not see the relationship to the size of the new collection.

Reply:
Are you sure?

consider a scenario where we configured `config.getParentsPerCommit() = 20.`

I have a branch1 (pointing to NO_ANCESTOR hash) and I am merging 3 commits to branch1. In that case `entry.getParents().size()` can only be 1 to 3, it can never be 20.

Reply:
Again: what is the relationship between the number of parents (in all cases) stored in a commit-entry and the number of unwritten entries? This definitely needs to be elaborated more in the comments. It's quite hard to guess what this code's really doing.

Reply:
Also, KeyList distance can be different than getParentsPerCommit.
So, I need to handle keylist size <, =, > parents size

and that key list size may not be pure unwritten, it can have both written and unwritten commits.

Reply:
Added comments and refactored a bit for better understanding.

Reply:
I'm not sure adding this complex logic is worth the benefit. Could we solve the problem by introducing a two-stage look up for commits involved in the key list construction: 1) from memory 2) if not found, from the database.

Callers can provide a real in-memory lookup function or a default no-op one.

We will not have to deal with all the complexities involved in configuring parent lists as longer or shorter than the "key distance". WDYT?

I'll try an open a PR for this quickly to illustrate my thinking.

Reply:
This is what I mean: #3486 if it looks ok in general, I will polish it of course and include feedback from this PR.

Reply:
sure. Go ahead. it looks good to me. I will close this PR.